File: packages/shared/src/match/stats.ts
export interface PlayerMatchStats {
  kills: number;
  headshots: number;
  shotsFired: number;
  hits: number;
  revives: number;
  downs: number;
  aliveMs: number;
  downedMs: number;
}

export function createPlayerMatchStats(): PlayerMatchStats {
  return {
    kills: 0,
    headshots: 0,
    shotsFired: 0,
    hits: 0,
    revives: 0,
    downs: 0,
    aliveMs: 0,
    downedMs: 0,
  };
}

export function resetPlayerMatchStats(stats: PlayerMatchStats): void {
  stats.kills = 0;
  stats.headshots = 0;
  stats.shotsFired = 0;
  stats.hits = 0;
  stats.revives = 0;
  stats.downs = 0;
  stats.aliveMs = 0;
  stats.downedMs = 0;
}

export function noteKill(stats: PlayerMatchStats, headshot: boolean): void {
  stats.kills += 1;
  if (headshot) stats.headshots += 1;
}

export function noteShot(stats: PlayerMatchStats): void {
  stats.shotsFired += 1;
}

export function noteHit(stats: PlayerMatchStats): void {
  stats.hits += 1;
}

export function noteDown(stats: PlayerMatchStats): void {
  stats.downs += 1;
}

export function noteRevive(stats: PlayerMatchStats): void {
  stats.revives += 1;
}

export function accumulateAlive(stats: PlayerMatchStats, dtMs: number): void {
  stats.aliveMs += dtMs;
}

export function accumulateDowned(stats: PlayerMatchStats, dtMs: number): void {
  stats.downedMs += dtMs;
}

export function accuracyOf(stats: PlayerMatchStats): number {
  return stats.hits / Math.max(1, stats.shotsFired);
}

export function survivalMsOf(stats: PlayerMatchStats, matchDurationMs: number): number {
  const survival = matchDurationMs - stats.downedMs;
  return survival > 0 ? survival : 0;
}

File: server/src/room/stats.cpp
#include "room/stats.hpp"

#include <cstring>

namespace ac::room {
namespace {

// 十进制写入 [first, last) 的低位在 last-1，返回新的 last。
char* writeDecimalBackwards(char* last, uint64_t value) noexcept {
  if (value == 0u) {
    *(--last) = '0';
    return last;
  }
  while (value > 0u) {
    *(--last) = static_cast<char>('0' + static_cast<char>(value % 10u));
    value /= 10u;
  }
  return last;
}

}  // namespace

void resetPlayerStats(PlayerStats& stats) noexcept { stats = PlayerStats{}; }

void noteKill(PlayerStats& stats, bool headshot) noexcept {
  stats.kills += 1u;
  if (headshot) stats.headshots += 1u;
}

void noteShot(PlayerStats& stats) noexcept { stats.shotsFired += 1u; }
void noteHit(PlayerStats& stats) noexcept { stats.hits += 1u; }
void noteDown(PlayerStats& stats) noexcept { stats.downs += 1u; }
void noteRevive(PlayerStats& stats) noexcept { stats.revives += 1u; }

void accumulateAlive(PlayerStats& stats, uint32_t dtMs) noexcept { stats.aliveMs += dtMs; }
void accumulateDowned(PlayerStats& stats, uint32_t dtMs) noexcept { stats.downedMs += dtMs; }

double accuracyOf(const PlayerStats& stats) noexcept {
  const uint32_t shots = stats.shotsFired > 0u ? stats.shotsFired : 1u;
  return static_cast<double>(stats.hits) / static_cast<double>(shots);
}

uint64_t survivalMsOf(const PlayerStats& stats, uint64_t matchDurationMs) noexcept {
  return matchDurationMs > stats.downedMs ? matchDurationMs - stats.downedMs : 0u;
}

std::size_t writeMatchId(char* out, std::size_t capacity, const char* roomCode,
                         uint64_t startedAtMs) noexcept {
  if (out == nullptr || capacity == 0u) return 0u;
  std::size_t used = 0u;
  if (roomCode != nullptr) {
    while (roomCode[used] != '\0' && used + 1u < capacity) {
      out[used] = roomCode[used];
      ++used;
    }
  }
  if (used + 1u < capacity) out[used++] = '-';
  char digits[24] = {};
  char* const begin = writeDecimalBackwards(digits + sizeof(digits), startedAtMs);
  const std::size_t count = static_cast<std::size_t>((digits + sizeof(digits)) - begin);
  for (std::size_t i = 0u; i < count && used + 1u < capacity; ++i) out[used++] = begin[i];
  out[used] = '\0';
  return used;
}

}  // namespace ac::room
